File: String-to-Integer(atoi)/String-to-Integer(atoi).cpp

class Solution {
public:
    int atoi(const char *str) {
        int start=0;
        while(str[start]==' ')
            start++;
        bool flag=true;
        if(str[start]=='+')
            start++;
        else if(str[start]=='-')
        {
            flag=false;
            start++;
        }
        int ans=0;
        for(int i=start;i<strlen(str);i++)
        {
            if(!isdigit(str[i]))
                break;
            if(ans>INT_MAX/10 || ans==INT_MAX/10 && str[i]-'0'>INT_MAX%10)
                if(flag)
                    return INT_MAX;
                else
                    return INT_MIN;
            ans=ans*10+str[i]-'0';
        }
        return flag?ans:-ans;
    }
};


 

File: String-to-Integer(atoi)/String-to-Integer(atoi).py


class Solution:
    # @return an integer
    def atoi(self, str):
        if str=="":
            return 0
        num=0
        flag=True
        n=len(str)
        start=0
        while str[start]==' ':
            start+=1
        if str[start]=='+':
            start+=1
        if str[start]=='-':
            start+=1
            flag=False
        for i in range(start,n):
            if str[i]<'0' or str[i]>'9':
                break
            if num>2147483647/10 or num==2147483647/10 and int(str[i])-int('0')>2147483647%10:
                if flag:
                    return 2147483647
                else:
                    return -2147483648
            num=num*10+int(str[i])-int('0')
        if flag:
            return num
        else:
            return -num
 